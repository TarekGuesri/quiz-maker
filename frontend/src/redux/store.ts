import { Action, ThunkAction, configureStore } from "@reduxjs/toolkit";

import ui from "./ui/ui-slice";

export const store = configureStore({
  reducer: {
    ui,
  },
  devTools: true,
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
