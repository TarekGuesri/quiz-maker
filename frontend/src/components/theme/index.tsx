import React, { FC } from 'react';
import CssBaseline from '@mui/material/CssBaseline';
import PropTypes from 'prop-types';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { useAppSelector } from 'src/redux/hooks';
import varelaRound from 'src/assets/fonts/VarelaRound-Regular.ttf';

export const Theme: FC = (props) => {
  const { darkMode } = useAppSelector((state) => state.ui);

  const theme = createTheme({
    palette: {
      mode: darkMode ? 'dark' : 'light',
      background: {
        default: darkMode ? '#333' : '#f7f7f7',
        paper: darkMode ? '#333' : '#fbfbfb',
      },
      primary: {
        main: '#8e49de',
        // contrastText:
      },
      secondary: {
        main: '#f9de4a',
      },
      contrastThreshold: 3,
      tonalOffset: 0.2,
    },
    typography: {
      fontFamily: 'Varela Round',
    },
    components: {
      MuiCssBaseline: {
        styleOverrides: `
          @font-face {
            font-family: 'Varela Round';
            font-style: normal;
            font-display: swap;
            font-weight: 400;
            src: url(${varelaRound}) format('truetype');
        `,
      },
    },
  });

  return (
    <ThemeProvider theme={theme}>
      {props.children}
      <CssBaseline />
    </ThemeProvider>
  );
};

Theme.propTypes = {
  children: PropTypes.node.isRequired,
};
