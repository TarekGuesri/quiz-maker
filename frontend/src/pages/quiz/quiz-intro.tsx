import React, { useState } from "react";
import { makeStyles } from "@mui/styles";
import Typography from "@mui/material/Typography";
import Button from "@mui/material/Button";
import { Theme } from "@mui/material";
import { store } from "src/redux/store";
import { startQuiz } from "src/redux/quiz/quiz-slice";
import ligthBulbImage from "src/assets/images/quiz-light-bulb.png";

const useStyles = makeStyles((theme: Theme) => ({
  ligthBulbImage: {
    width: "75px",
    [theme.breakpoints.down("md")]: {
      width: "50px",
    },
  },
  ligthBulbBackground: {
    border: `9px solid ${theme.palette.secondary.main}`,
    backgroundColor: `${theme.palette.primary.main}`,
    display: "inline-flex",
    padding: "18px",
    borderRadius: "100px",
    marginBottom: theme.spacing(4),
  },
  button: {
    width: "150px",
    padding: "10px",
  },
}));

interface QuizIntroProps {
  description?: string | undefined;
}

export const QuizIntro: React.FC<QuizIntroProps> = ({ description }) => {
  const classes = useStyles();
  // We use this to hide div style (border and backround) when image is loading
  const [imageLoading, setImageLoading] = useState(true);

  return (
    <>
      <div className={imageLoading ? "" : classes.ligthBulbBackground}>
        <img
          src={ligthBulbImage}
          alt="light-bulb-image"
          className={classes.ligthBulbImage}
          onLoad={() => setImageLoading(false)}
          data-testid="light-bulb-image"
        />
      </div>

      <Typography mb={4} data-testid="quiz-description">
        {description || "This quiz doesn't have a descirption."}
      </Typography>

      <Button
        variant="contained"
        color="primary"
        className={classes.button}
        onClick={() => store.dispatch(startQuiz())}
      >
        Start Quiz
      </Button>
    </>
  );
};
