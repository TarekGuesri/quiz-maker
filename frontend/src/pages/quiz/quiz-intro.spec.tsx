import React from "react";
import { cleanup, render, screen, fireEvent } from "@testing-library/react";
import { store } from "src/redux/store";
import { Provider } from "react-redux";
import { Theme } from "src/components/theme";
import { QuizIntro } from "./quiz-intro";

describe("src/pages/quiz/quiz-intro.tsx", () => {
  afterEach(() => cleanup());

  test("Render QuizIntro component", () => {
    const renderedComponent = render(
      <Provider store={store}>
        <Theme>
          <QuizIntro description="Description test" />
        </Theme>
      </Provider>,
    );

    // We wait for the image to load
    const ligthBulbImage = screen.getByTestId("light-bulb-image");
    fireEvent.load(ligthBulbImage);

    expect(renderedComponent.container).toMatchSnapshot();
  });

  test("when QuizIntro doesn't have a description, we display 'This quiz doesn't have a descirption.'", () => {
    render(
      <Provider store={store}>
        <Theme>
          <QuizIntro />
        </Theme>
      </Provider>,
    );

    const text = screen.getByTestId("quiz-description");

    expect(text).toHaveTextContent("This quiz doesn't have a descirption.");
  });

  test("when pressing Start Quiz button, quiz should start", () => {
    render(
      <Provider store={store}>
        <Theme>
          <QuizIntro />
        </Theme>
      </Provider>,
    );

    expect(store.getState().quiz.quizStarted).toBeFalsy();

    const startButton = screen.getByText(/start quiz/i);
    fireEvent.click(startButton);

    expect(store.getState().quiz.quizStarted).toBeTruthy();
  });
});
