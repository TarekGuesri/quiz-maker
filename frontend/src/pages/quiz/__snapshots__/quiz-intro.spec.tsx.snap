// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/quiz/quiz-intro.tsx Render QuizIntro component 1`] = `
<div>
  <div
    class="makeStyles-ligthBulbBackground-2"
  >
    <img
      alt="light-bulb-image"
      class="makeStyles-ligthBulbImage-1"
      data-testid="light-bulb-image"
      src="quiz-light-bulb.png"
    />
  </div>
  <p
    class="MuiTypography-root MuiTypography-body1 css-ig1lqu-MuiTypography-root"
    data-testid="quiz-description"
  >
    Description test
  </p>
  <button
    class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root makeStyles-button-3 css-kygxw5-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Start Quiz
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
